sys/build.rs

```diff
@@ -129,21 +127,57 @@ fn feature_to_define(name: impl AsRef<str>) -> String {
 }
 
 fn patch<D: AsRef<Path>, P: AsRef<Path>>(out_dir: D, patch: P) {
-    let mut child = Command::new("patch")
-        .arg("-p1")
-        .stdin(Stdio::piped())
-        .current_dir(out_dir)
-        .spawn()
-        .expect("Unable to execute patch, you may need to install it: {}");
+    struct Patches<'a>(&'a str);
+
+    // A backtracking attempt to find a valid diff when there are multiple patches in one file (also known as patchset)
+    impl<'a> Iterator for Patches<'a> {
+        type Item = Patch<'a, str>;
+        fn next(&mut self) -> Option<Self::Item> {
+            let mut range = self.0.len();
+            loop {
+                match Patch::from_str(&self.0[..range]) {
+                    Ok(x) if x.hunks().is_empty() => break None,
+                    Err(_) if range < 1 => break None,
+                    Ok(x) => {
+                        self.0 = &self.0[range..];
+                        break Some(x);
+                    }
+                    Err(_) => range -= 1,
+                }
+            }
+        }
+    }
+
     println!("Appliyng patch {}", patch.as_ref().display());
     {
-        let patch = fs::read(patch).expect("Unable to read patch");
-
-        let stdin = child.stdin.as_mut().unwrap();
-        stdin.write_all(&patch).expect("Unable to apply patch");
+        let out_dir = out_dir.as_ref();
+        let patch = fs::read_to_string(patch).expect("Unable to read patch");
+        let patch = dos2unix(&patch);
+        for patch in Patches(&patch) {
+            let original = patch
+                .original()
+                .and_then(|x| x.split_once('/').map(|(_, b)| b))
+                .or(patch.original())
+                .expect("Cannot find original file name");
+            let modified = patch
+                .modified()
+                .and_then(|x| x.split_once('/').map(|(_, b)| b))
+                .or(patch.modified())
+                .expect("Cannot find modified file name");
+
+            // for now
+            assert_eq!(original, modified);
+
+            let original =
+                fs::read_to_string(out_dir.join(original)).expect("Unable to read original file");
+            let original = dos2unix(&original);
+            match apply(&original, &patch) {
```

Comment:
Does it support fuzzy patch applying?
Could you test build with `parallel` feature on windows?

Reply:
Not sure about fuzzy patching, but I am building it with parallel and futures

Reply:
The fact is that MSVC patches changes line numbers and may interfere applying other patches.